File: quant1x/id/Id.java
// Copyright (c) 2026 Quant1X. All rights reserved.
// Author: wangfeng <wangfengxy@sina.cn>
// SPDX-License-Identifier: MIT

package quant1x.id;

import java.util.Arrays;
import java.util.Base64;

/**
 * ID 的解析视图：从 16 字节原始数据中提取 hlc / nodeID / seq，
 * 并提供 base64url（无填充）字符串与字节数组两种交换格式，
 * 与 Go 版的 {@code ID} / Python 版 {@code id} 对齐。
 */
public final class Id {

    private final byte[] raw;

    private Id(byte[] raw) {
        if (raw.length != 16) {
            throw new IllegalArgumentException("ID expects exactly 16 bytes, got " + raw.length);
        }
        this.raw = raw.clone();
    }

    /**
     * 由 128 位整数构造 ID。
     */
    public static Id fromUint128(Uint128 value) {
        return new Id(value.toBytes());
    }

    /**
     * 由 16 字节原始数据构造 ID。
     *
     * @throws IllegalArgumentException 当 {@code raw.length != 16}
     */
    public static Id fromBytes(byte[] raw) {
        return new Id(raw);
    }

    /**
     * 原始 16 字节（BigEndian）。
     */
    public byte[] bytes() {
        return raw.clone();
    }

    /**
     * base64url 无填充字符串（与 Go 的 RawURLEncoding / Python 的
     * {@code urlsafe_b64encode().rstrip(b"=")} 一致）。
     */
    @Override
    public String toString() {
        return Base64.getUrlEncoder().withoutPadding().encodeToString(raw);
    }

    /** hlc（高 64 位） */
    public long hlc() {
        return Uint128.readUint64(raw, 0);
    }

    /** nodeID（32 位，以 long 承载无符号值） */
    public long nodeId() {
        return Uint128.readUint32(raw, 8);
    }

    /** seq（32 位，以 long 承载无符号值） */
    public long seq() {
        return Uint128.readUint32(raw, 12);
    }

    @Override
    public boolean equals(Object obj) {
        if (this == obj) {
            return true;
        }
        if (!(obj instanceof Id)) {
            return false;
        }
        return Arrays.equals(raw, ((Id) obj).raw);
    }

    @Override
    public int hashCode() {
        return Arrays.hashCode(raw);
    }
}

File: quant1x/id/id.go
package id

import (
	"encoding/base64"
	"encoding/binary"
)

// ID 是按原始字节编码的 128 位可排序标识。
// 它可比较，也可以直接作为 map 键使用。
type ID [16]byte

// Bytes 返回原始字节切片。
func (id ID) Bytes() []byte {
	return id[:]
}

// String 返回不带填充的 base64url 字符串。
func (id ID) String() string {
	return base64.RawURLEncoding.EncodeToString(id[:])
}

// NodeID 提取节点标识。
func (id ID) NodeID() uint32 {
	return binary.BigEndian.Uint32(id[8:12])
}

// Seq 提取 32 位序列号。
func (id ID) Seq() uint32 {
	return binary.BigEndian.Uint32(id[12:16])
}

// HLC 提取原始 HLC 高位时间戳。
func (id ID) HLC() uint64 {
	return binary.BigEndian.Uint64(id[0:8])
}
